libkbfs/journal_md_ops.go

```diff
@@ -0,0 +1,382 @@
+	// Need to always consult the server to get the tlfID. No need
+	// to optimize this, since all subsequent lookups will be by
+	// TLF. Although if we did want to, we could store a handle ->
+	// TLF ID mapping with the journals.
+	tlfID, rmd, err := j.MDOps.GetForHandle(ctx, handle, mStatus)
+	if err != nil {
+		return TlfID{}, ImmutableRootMetadata{}, err
+	}
+
+	lookupTlfID := tlfID
+	if rmd != (ImmutableRootMetadata{}) {
+		lookupTlfID = rmd.ID
+	}
+
+	// If the journal has a head, use that.
+	irmd, err := j.getHeadFromJournal(
+		ctx, lookupTlfID, NullBranchID, mStatus, handle)
+	if err != nil {
+		return TlfID{}, ImmutableRootMetadata{}, err
+	}
+	if irmd != (ImmutableRootMetadata{}) {
+		return TlfID{}, irmd, nil
+	}
+
+	// Otherwise, use the server's head.
+	return tlfID, rmd, nil
+}
+
+// TODO: Combine the two GetForTLF functions in MDOps to avoid the
+// need for these helper functions.
+func (j journalMDOps) getForTLF(
+	ctx context.Context, id TlfID, bid BranchID, mStatus MergeStatus,
+	delegateFn func(context.Context, TlfID) (ImmutableRootMetadata, error)) (
+	ImmutableRootMetadata, error) {
+	// If the journal has a head, use that.
+	irmd, err := j.getHeadFromJournal(ctx, id, bid, mStatus, nil)
+	if err != nil {
+		return ImmutableRootMetadata{}, err
+	}
+	if irmd != (ImmutableRootMetadata{}) {
+		return irmd, nil
+	}
+
+	// Otherwise, consult the server instead.
+	return delegateFn(ctx, id)
+}
+
+func (j journalMDOps) GetForTLF(
+	ctx context.Context, id TlfID) (ImmutableRootMetadata, error) {
+	return j.getForTLF(ctx, id, NullBranchID, Merged, j.MDOps.GetForTLF)
+}
+
+func (j journalMDOps) GetUnmergedForTLF(
+	ctx context.Context, id TlfID, bid BranchID) (
+	ImmutableRootMetadata, error) {
+	delegateFn := func(ctx context.Context, id TlfID) (
+		ImmutableRootMetadata, error) {
+		return j.MDOps.GetUnmergedForTLF(ctx, id, bid)
+	}
+	return j.getForTLF(ctx, id, bid, Unmerged, delegateFn)
+}
+
+// TODO: Combine the two GetForTLF functions in MDOps to avoid the
+// need for these helper functions.
+func (j journalMDOps) getRange(
+	ctx context.Context, id TlfID, bid BranchID, mStatus MergeStatus,
+	start, stop MetadataRevision,
+	delegateFn func(ctx context.Context, id TlfID,
+		start, stop MetadataRevision) (
+		[]ImmutableRootMetadata, error)) (
+	[]ImmutableRootMetadata, error) {
+	// Grab the range from the journal first.
+	jirmds, err := j.getRangeFromJournal(ctx, id, bid, mStatus, start, stop)
+	if err != nil {
+		return nil, err
+	}
+
+	// If it's empty, just fall back to the server.
+	if len(jirmds) == 0 {
+		return delegateFn(ctx, id, start, stop)
+	}
+
+	// If the first revision from the journal is the first
+	// revision we asked for, then just return the range from the
+	// journal.
+	if jirmds[0].Revision == start {
+		return jirmds, nil
+	}
+
+	// Otherwise, fetch the rest from the server and prepend them.
+	serverStop := jirmds[0].Revision - 1
+	irmds, err := delegateFn(ctx, id, start, stop)
```

Comment:
D'oh, nice catch. I'm punting on more tests for `journalMdOps` (and to clean up the existing one) for a future PR, but added a TODO.